src/loader/implementation.ml

```diff
@@ -4,367 +4,182 @@
 
 let pos_of_loc loc = (loc.Location.loc_start.pos_cnum, loc.loc_end.pos_cnum)
 
-type annotations =
-  | LocalDefinition of Ident.t
-  | LocalValue of Ident.t
-  | DefJmp of Shape.Uid.t
-
 let counter =
   let c = ref 0 in
   fun () ->
     incr c;
     !c
 
-module Analysis = struct
+module Env = struct
   open Typedtree
   open Odoc_model.Paths
 
-  type env = Ident_env.t * Location.t Shape.Uid.Tbl.t
-
-  let env_wrap : (Ident_env.t -> Ident_env.t) -> env -> env =
-   fun f (env, uid_to_loc) -> (f env, uid_to_loc)
-
-  let get_env : env -> Ident_env.t = fun (env, _) -> env
+  let rec structure env parent str =
+    let env' = Ident_env.add_structure_tree_items parent str env in
+    List.iter (structure_item env' parent) str.str_items
 
-  let get_uid_to_loc : env -> Location.t Shape.Uid.Tbl.t =
-   fun (_, uid_to_loc) -> uid_to_loc
+  and signature env parent sg =
+    let env' = Ident_env.add_signature_tree_items parent sg env in
+    List.iter (signature_item env' parent) sg.sig_items
 
-  let rec structure env parent acc str =
-    let env' = env_wrap (Ident_env.add_structure_tree_items parent str) env in
-    List.fold_left (structure_item env' parent) acc str.str_items
-
-  and signature env parent acc sg =
-    let env' = env_wrap (Ident_env.add_signature_tree_items parent sg) env in
-    List.fold_left (signature_item env' parent) acc sg.sig_items
-
-  and signature_item env parent acc item =
+  and signature_item env parent item =
     match item.sig_desc with
-    | Tsig_value vd -> value_description env parent acc vd
-    | Tsig_type (_, tds) -> type_declarations env parent acc tds
-    | Tsig_typesubst tds -> type_declarations env parent acc tds
-    | Tsig_typext _ -> acc
-    | Tsig_exception e -> exception_ env parent acc e
-    | Tsig_module mb -> module_declaration env parent acc mb
-    | Tsig_modsubst ms -> module_substitution env parent acc ms
-    | Tsig_recmodule mbs -> module_declarations env parent acc mbs
-    | Tsig_modtype mtd -> module_type_declaration env parent acc mtd
-    | Tsig_modtypesubst mtd -> module_type_declaration env parent acc mtd
-    | Tsig_open _ -> acc
-    | Tsig_include _ -> acc
-    | Tsig_class cd -> class_description env parent acc cd
-    | Tsig_class_type ctd -> class_type_declaration env parent acc ctd
-    | Tsig_attribute _ -> acc
-
-  and value_description _env _parent acc _vd = acc
-
-  (* and type_declaration _env _parent _td = [] *)
-
-  and type_declarations _env _parent acc _tds = acc
-
-  and exception_ _env _parent acc _e = acc
-
-  and module_declaration env _parent acc md =
+    | Tsig_module mb -> module_declaration env parent mb
+    | Tsig_recmodule mbs -> module_declarations env parent mbs
+    | Tsig_modtype mtd -> module_type_declaration env parent mtd
+    | Tsig_modtypesubst mtd -> module_type_declaration env parent mtd
+    | Tsig_value _ | Tsig_type _ | Tsig_typesubst _ | Tsig_typext _
+    | Tsig_exception _ | Tsig_modsubst _ | Tsig_open _ | Tsig_include _
```

Comment:
Why is `Tsig_modsubst` not iterated over ?

Reply:
That's a good question. I'm uncomfortable writing all those traverse by hand, since it's easy to miss something, particularly on the language constructs I don't know so well. I'll check and add the traverse if needed!

Reply:
It is not iterated because there is nothing to iterate inside!
```
and module_substitution =
    {
     ms_id: Ident.t;
     ms_name: string loc;
     ms_manifest: Path.t;
     ms_txt: Longident.t loc;
     ms_attributes: attributes;
     ms_loc: Location.t;
    }
```
In the parsetree, it is used to represent
```
  | Psig_modsubst of module_substitution  (** [module X := M] *)
```
and I assume it represents the same in the typedtree.

Reply:
I see! It must be added to the env but that's already done in `Ident_env`, so there's no problem. Sorry for the false alarm.